scriptworker/test/test_cot_verify.py

```diff
@@ -2367,3 +2367,64 @@ def cot(*args, **kwargs):
     mocker.patch.object(cotverify, 'verify_chain_of_trust', new=noop_async)
 
     cotverify.verify_cot_cmdln(args=args, event_loop=event_loop)
+
+
+# create_test_workdir {{{1
+@pytest.mark.asyncio
+async def test_async_create_test_workdir(mocker, tmpdir):
+    """``_async_create_test_workdir`` populates ``path`` with ``task.json`` and
+    cot artifacts.
+
+    """
+    task_defn = {
+        "payload": {
+            "upstreamArtifacts": [{
+                "taskId": "taskId1",
+                "paths": ["public/build/one", "public/build/two"],
+            }],
+        },
+    }
+
+    async def fake_task(task_id):
+        return task_defn
+
+    queue = mocker.MagicMock()
+    queue.task = fake_task
+    mocker.patch.object(cotverify, "download_artifacts", new=noop_async)
+
+    await cotverify._async_create_test_workdir("taskId", tmpdir, queue=queue)
+    contents = load_json_or_yaml(os.path.join(tmpdir, "task.json"), is_path=True)
+    assert contents == task_defn
+
+
+@pytest.mark.parametrize("exists, overwrite, raises", ((
+    True, True, False,
+), (
+    False, False, False,
+), (
+    False, True, False,
+), (
+    True, False, True,
+)))
+def test_create_test_workdir(mocker, event_loop, tmpdir, exists, overwrite, raises):
+    """``create_test_workdir`` fails if ``path`` exists and ``--overwrite`` isn't
+    passed. Otherwise we call ``_async_create_test_workdir`` with the appropriate
+    args.
+
+    """
+    work_dir = os.path.join(tmpdir, "work")
+    args = ["--path", work_dir, "taskId"]
+    if overwrite:
+        args.append("--overwrite")
+
+    async def fake_create(*args):
+        assert args == ("taskId", work_dir)
+
+    mocker.patch.object(cotverify, "_async_create_test_workdir", new=fake_create)
+    if exists:
+        makedirs(work_dir)
+    if raises:
+        with pytest.raises(SystemExit):
```

Comment:
FWIW, To make it easier to read the flow of this test, you could break it into four simpler tests :thinking:

Reply:
Sure. That would make it easier to read each test, by requiring more effort to create each test. I tend to rely on parameters to help test multiple code branches. The theory is if we make testing code branches easier to write, people will write more tests.

Reply:
I hear you, though of course sometimes it's not just the amount of code to write that contributes to the friction of writing tests. If I attempt to capture additional test cases, but existing parameter sets fail in unexpected ways (or if it's difficult to imagine how new tests would fit into the existing generic test structure), the costs might outweigh the theoretical benefit of fewer lines of test code needing to be written.

Reply:
We have the pattern of the `raises` parameter all over the scriptworker tests, so changing that behavior here might not be a huge win overall. I could split this into two functions, one with a `makedirs` and one without. My first instinct would be to create a helper function that does everything `test_create_test_workdir` currently does, and create stub test functions that call the helper functions, but without parameterizing, but I don't think that helps readability at all.

Sounds like you consider this test change a blocker?

Reply:
Hmm, interesting :thinking: 
It is true that we have a convention of the `raises` parameter elsewhere, too. I'm definitely not worried about this, it _isn't_ a blocker. Feel free to merge! I think this is an interesting discussion either way :)

I agree that just moving the parameters into separate tests but still having a generic, separate `test_create_test_workdir` function wouldn't improve readability.
I'd probably lean on duplication and create four very straightforward tests:

```
async def fake_create_test_workdir(*args):
    assert args == ("taskId", work_dir)


def test_create_test_workdir(mocker, event_loop, tmpdir):
    work_dir = os.path.join(tmpdir, "work")
    mocker.patch.object(cotverify, "_async_create_test_workdir", new=fake_create_test_workdir)
    makedirs(work_dir)
    cotverify.create_test_workdir(args=["--path", work_dir, "taskId", "--overwrite"], event_loop=event_loop)


def test_create_test_workdir(mocker, event_loop, tmpdir):
    work_dir = os.path.join(tmpdir, "work")
    mocker.patch.object(cotverify, "_async_create_test_workdir", new=fake_create_test_workdir)
    cotverify.create_test_workdir(args=["--path", work_dir, "taskId"], event_loop=event_loop)

def test_create_test_workdir(mocker, event_loop, tmpdir):
    work_dir = os.path.join(tmpdir, "work")
    mocker.patch.object(cotverify, "_async_create_test_workdir", new=fake_create_test_workdir)
    cotverify.create_test_workdir(args=["--path", work_dir, "taskId", "--overwrite"], event_loop=event_loop)


def test_create_test_workdir(mocker, event_loop, tmpdir):
    work_dir = os.path.join(tmpdir, "work")
    mocker.patch.object(cotverify, "_async_create_test_workdir", new=fake_create_test_workdir)
    makedirs(work_dir)
    with pytest.raises(SystemExit):
        cotverify.create_test_workdir(args=["--path", work_dir, "taskId"], event_loop=event_loop)
```

Reply:
I agree each individual test is easier to read in your example. We'd need to avoid duplicate test names, so we could add something like `_no_existing_work_dir_no_overwrite` to the test name.

Later on, if we refactored `create_test_workdir` significantly, it might mean we need to edit 4 test functions rather than 1. It might not be immediately clear which tests should stick around (which tests are still useful) and which are essentially duplicates of other tests. Also, let's say we change the behavior of the code that affects a subset of the tests: we'd need to figure out which test(s) to change. This could be 1, 4, or if we take this to more of an extreme, 10 or 100. Also, if we have enough of these functions with only minor changes between them, it may end up being a "where's waldo" or "can you find the 10 differences between these images" type game figuring out what's different between them, especially if the functions grow to more than 3-5 lines. Alternately if we have descriptive enough test names, the function names may approach the code length of the simpler tests ;-)

I don't think either approach is perfect for all cases. I still prefer the parameterization approach for now, so I'll merge. Thanks!

Reply:
> Later on, if we refactored create_test_workdir significantly, it might mean we need to edit 4 test functions rather than 1. It might not be immediately clear which tests should stick around (which tests are still useful) and which are essentially duplicates of other tests.

It's true that more test functions would need to be edited, though fortunately, it's work that's hard to do incorrectly, since the tests won't (shouldn't :stuck_out_tongue:) pass until they've all been updated.

However, and more importantly, is knowing which tests become obsolete. With a single generic parameterized test, wouldn't it be _more_ challenging to realize which parameter sets are obsolete?

> This could be 1, 4, or if we take this to more of an extreme, 10 or 100.

Of course, as the number of tests scale, when you find behaviour duplication within your tests, you should absolutely apply re-use accordingly. IMHO, the abstraction of re-use applied in this particular example is less valuable.

-----

Reading over your comment again, I think we're pretty aligned, but we draw the line of "at what point should code be re-used" at different parts of the scale. :+1:

Reply:
> Reading over your comment again, I think we're pretty aligned, but we draw the line of "at what point should code be re-used" at different parts of the scale. 👍

I agree with this a lot. And I could be convinced that for super simple tests like this one, a different approach than parameterization might be better. We use parameterization across most (all?) the scriptworker tests, though, and I don't love having to rewrite simple tests to a parametrized test once the function passes a certain level of complexity. I think consistently using parameters across scriptworker tests is worth the readability costs of parameterized simpler tests.

Reply:
> I don't love having to rewrite [a set of multiple] simple tests to a [single] parameterized test once the function passes a certain level of complexity. I think consistently using parameters across scriptworker tests is worth the readability costs of parameterized simpler tests.

I agree that as we add more tests, our desire to add re-use between tests will definitely grow. I'm not sure that that re-use will always manifest by condensing multiple tests into a single, well-parameterized test, though I'm sure that it will still be sometimes applicable.

One other interesting observation is how the re-use is applied - I feel that it can be more straightforward to have "bottom-up" re-use in tests (share common grouped code like `fake_create_test_workdir`) rather than "top-down" (share a full test workflow, with different branches enabled/disabled with parameters)